File: frontend/src/lib/schemas/actuals.js
/**
 * Zod schemas for the Actuals API (Chat 19B §R1.1).
 *
 * Mirror the backend `_serialise_actual` response. Sensitive fields are
 * `.nullable().optional()` — a single schema parses both gated (full) and
 * ungated (stripped) responses (chat-17 E7 pattern). UI checks
 * `=== undefined` AND `=== null` and renders "—" in both cases. The
 * existing `lib/format.js::formatMoney(v)` already handles both — pass it
 * `null | undefined | "1234.56"` and it returns "—" / "—" / "£1,234.56".
 *
 * Decimals come over the wire as strings (backend Decimal → str). The
 * schema validates the pattern; render code converts only on display.
 */
import { z } from 'zod';

// Decimal-as-string for money. Backend serialises Decimal to string to avoid
// JS float drift. Up to 2 decimal places.
const moneyString = z
  .string()
  .regex(/^-?\d+(\.\d{1,2})?$/, 'money must be a decimal string with up to 2dp')
  .nullable();

const uuid = z.string().uuid();
const isoDateOrNull = z.string().regex(/^\d{4}-\d{2}-\d{2}$/).nullable();
const isoDateTimeOrNull = z.string().datetime({ offset: true }).nullable();

// ─── Enums ───────────────────────────────────────────────────────────
export const ActualStatusEnum = z.enum([
  'Draft', 'Posted', 'Paid', 'Disputed', 'Void',
]);

export const ActualSourceTypeEnum = z.enum([
  'Manual_Entry', 'Xero_Bill', 'Xero_Credit_Note', 'SC_Valuation',
  'Day_Rate_Timesheet', 'Expense_Claim', 'Journal', 'Internal_Recharge',
]);

// ─── Model ───────────────────────────────────────────────────────────
export const ActualSchema = z.object({
  id: uuid,
  project_id: uuid,
  budget_line_id: uuid,
  entity_id: uuid,
  source_type: ActualSourceTypeEnum,
  source_reference: z.string().nullable(),
  external_id: z.string().nullable(),
  transaction_date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/),
  posting_date: isoDateOrNull,
  description: z.string(),
  net_amount: moneyString,
  vat_amount: moneyString,
  gross_amount: moneyString,
  // Rates can be 3dp (e.g. CIS 12.345%) — no 2dp constraint.
  vat_rate_pct: z.string().nullable(),
  is_vat_recoverable: z.boolean(),
  currency: z.string().length(3),
  exchange_rate: z.string().nullable(),
  supplier_id: uuid.nullable(),
  supplier_name_snapshot: z.string(),
  supplier_invoice_ref: z.string().nullable(),
  is_cis_applicable: z.boolean(),
  retention_released: z.boolean(),
  linked_commitment_id: uuid.nullable(),
  related_subcontract_id: uuid.nullable(),
  is_reconciled_to_xero: z.boolean(),
  status: ActualStatusEnum,
  posted_at: isoDateTimeOrNull,
  paid_date: isoDateOrNull,
  payment_reference: z.string().nullable(),
  disputed_at: isoDateTimeOrNull,
  voided_at: isoDateTimeOrNull,
  created_at: isoDateTimeOrNull,
  updated_at: isoDateTimeOrNull,
  // Sensitive — present only when caller has actuals.view_sensitive
  cis_deduction_rate_pct: z.string().nullable().optional(),
  cis_labour_amount: moneyString.optional(),
  cis_materials_amount: moneyString.optional(),
  cis_deduction_amount: moneyString.optional(),
  retention_rate_pct: z.string().nullable().optional(),
  retention_amount: moneyString.optional(),
  retention_release_date: isoDateOrNull.optional(),
  reconciliation_variance: moneyString.optional(),
  dispute_reason: z.string().nullable().optional(),
  void_reason: z.string().nullable().optional(),
  ai_capture_metadata: z.record(z.unknown()).nullable().optional(),
});

export const ActualListResponseSchema = z.object({
  items: z.array(ActualSchema),
  count: z.number().int(),
  total: z.number().int(),
  limit: z.number().int().optional(),
  offset: z.number().int().optional(),
  // Present on /projects/:id/actuals only.
  project_id: uuid.optional(),
});

// ─── Change log ──────────────────────────────────────────────────────
export const ChangeLogEventSchema = z.object({
  id: uuid,
  actual_id: uuid,
  event_type: z.string(),
  actor_user_id: uuid.nullable(),
  event_payload: z.record(z.unknown()).nullable(),
  occurred_at: isoDateTimeOrNull,
});

export const ChangeLogResponseSchema = z.object({
  actual_id: uuid,
  items: z.array(ChangeLogEventSchema),
  count: z.number().int(),
});

// ─── Attachments ─────────────────────────────────────────────────────
export const AttachmentSchema = z.object({
  id: uuid,
  actual_id: uuid,
  original_filename: z.string(),
  file_type: z.string(),
  file_size_bytes: z.number().int(),
  source: z.enum(['Manual_Upload', 'Email_Capture', 'AI_Capture']),
  uploaded_by_user_id: uuid.nullable(),
  uploaded_at: isoDateTimeOrNull,
});

// ─── Request schemas — mirror backend Pydantic strictly ──────────────
export const CreateActualRequestSchema = z.object({
  project_id: uuid,
  budget_line_id: uuid,
  entity_id: uuid,
  source_type: ActualSourceTypeEnum,
  source_reference: z.string().max(10000).optional(),
  external_id: z.string().max(255).optional(),
  transaction_date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/),
  posting_date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/).optional(),
  description: z.string().min(1).max(10000),
  net_amount: z.string().regex(/^-?\d+(\.\d{1,2})?$/),
  vat_amount: z.string().regex(/^-?\d+(\.\d{1,2})?$/).default('0'),
  vat_rate_pct: z.string().regex(/^-?\d+(\.\d{1,3})?$/).default('20'),
  is_vat_recoverable: z.boolean().default(true),
  currency: z.string().length(3).default('GBP'),
  exchange_rate: z.string().optional(),
  supplier_id: uuid.optional(),
  supplier_name_snapshot: z.string().min(1).max(255),
  supplier_invoice_ref: z.string().max(100).optional(),
  is_cis_applicable: z.boolean().default(false),
  cis_deduction_rate_pct: z.string().optional(),
  cis_labour_amount: z.string().optional(),
  cis_materials_amount: z.string().optional(),
  retention_rate_pct: z.string().optional(),
  retention_amount: z.string().optional(),
  linked_commitment_id: uuid.optional(),
  related_subcontract_id: uuid.optional(),
});

export const MarkPaidRequestSchema = z.object({
  paid_date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/),
  payment_reference: z.string().min(1).max(255),
});

export const VoidActualRequestSchema = z.object({
  void_reason: z.string().min(3).max(10000),
});

export const DisputeActualRequestSchema = z.object({
  dispute_reason: z.string().min(3).max(10000),
});

export const UndisputeActualRequestSchema = z.object({
  notes: z.string().max(10000).optional(),
});

export const ReleaseRetentionRequestSchema = z.object({
  retention_release_date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/),
});

File: backend/app/routers/actuals.py
"""Actuals API — Prompt 2.5A / Chat 19A.

Mounts under /api/v1. 15 endpoints:

  Project-scoped list:
    GET    /projects/{project_id}/actuals                     list actuals

  Generic list + create:
    GET    /actuals                                           list (filters)
    POST   /actuals                                           create Draft   [actuals.create]

  Single actual:
    GET    /actuals/{actual_id}                               detail
    PATCH  /actuals/{actual_id}                               edit Draft     [actuals.edit]
    DELETE /actuals/{actual_id}                               delete Draft   [actuals.edit]
    GET    /actuals/{actual_id}/change-log                    change history

  State transitions:
    POST   /actuals/{actual_id}/post                          Draft → Posted [actuals.post]
    POST   /actuals/{actual_id}/mark-paid                     Posted → Paid  [actuals.approve]
    POST   /actuals/{actual_id}/void                          → Void         [actuals.approve]
    POST   /actuals/{actual_id}/dispute                       Posted → Disputed [actuals.edit]
    POST   /actuals/{actual_id}/undispute                     Disputed → Posted [actuals.edit]
    POST   /actuals/{actual_id}/release-retention             retention release [actuals.approve]

  Attachments:
    GET    /actuals/{actual_id}/attachments                   list
    POST   /actuals/{actual_id}/attachments                   multipart upload [actuals.edit]
    DELETE /actuals/{actual_id}/attachments/{att_id}          remove         [actuals.edit]

Tenant scoping: Pattern α via _visible_project_ids (same as budgets/appraisals).
"""
from __future__ import annotations

import uuid
from typing import Any, Optional

from fastapi import (
    APIRouter, Depends, File, Form, HTTPException, Query, Request, UploadFile,
    Response,
)
from pydantic import ValidationError
from sqlalchemy.orm import Session

from app.auth.deps import get_current_user, require_permission
from app.auth.permissions import UserPermissions
from app.db import get_db
from app.models.actuals import Actual, ActualAttachment, ActualChangeLog
from app.models.user import User
from app.schemas.actuals import (
    ActualsListFilters, CreateActualRequest, DisputeActualRequest,
    MarkPaidRequest, PostActualRequest, ReleaseRetentionRequest,
    UndisputeActualRequest, UpdateDraftActualRequest, VoidActualRequest,
)
from app.services import actuals as actuals_svc
from app.services import actual_attachments as att_svc
from app.services.actual_errors import ActualError


router = APIRouter(tags=["actuals"])


def _actuals_filters_dep(
    project_id: Optional[uuid.UUID] = Query(default=None),
    budget_line_id: Optional[uuid.UUID] = Query(default=None),
    entity_id: Optional[uuid.UUID] = Query(default=None),
    status: Optional[str] = Query(default=None),
    source_type: Optional[str] = Query(default=None),
    supplier_id: Optional[uuid.UUID] = Query(default=None),
    transaction_date_from: Optional[str] = Query(default=None),
    transaction_date_to: Optional[str] = Query(default=None),
    limit: int = Query(default=100, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
) -> ActualsListFilters:
    """D32 — wraps `ActualsListFilters` construction so that a
    `ValidationError` raised from a `@field_validator` (e.g. comma-separated
    status containing an invalid value) is converted to a clean
    `HTTPException(422)` instead of escaping the dependency stack and
    surfacing as a 500.
    """
    try:
        return ActualsListFilters(
            project_id=project_id,
            budget_line_id=budget_line_id,
            entity_id=entity_id,
            status=status,
            source_type=source_type,
            supplier_id=supplier_id,
            transaction_date_from=transaction_date_from,
            transaction_date_to=transaction_date_to,
            limit=limit,
            offset=offset,
        )
    except ValidationError as exc:
        raise HTTPException(
            status_code=422,
            detail=exc.errors(include_url=False, include_context=False),
        ) from exc


# ---------------------------------------------------------------------
# Error mapping
# ---------------------------------------------------------------------

def _raise_for(exc: ActualError):
    raise HTTPException(
        exc.http_status,
        detail={"code": exc.code, "message": str(exc), "details": exc.details},
    )


# ---------------------------------------------------------------------
# Serialisers
# ---------------------------------------------------------------------

def _serialise_actual(a: Actual, *, include_sensitive: bool) -> dict:
    """Build the response shape. `include_sensitive` gates retention/CIS detail.

    The retention/CIS *header* fields are always shown so users see status;
    `include_sensitive` adds the calculated amounts.
    """
    base: dict[str, Any] = {
        "id": str(a.id),
        "project_id": str(a.project_id),
        "budget_line_id": str(a.budget_line_id),
        "entity_id": str(a.entity_id),
        "source_type": a.source_type,
        "source_reference": a.source_reference,
        "external_id": a.external_id,
        "transaction_date": a.transaction_date.isoformat() if a.transaction_date else None,
        "posting_date": a.posting_date.isoformat() if a.posting_date else None,
        "description": a.description,
        "net_amount": str(a.net_amount),
        "vat_amount": str(a.vat_amount),
        "gross_amount": str(a.gross_amount),
        "vat_rate_pct": str(a.vat_rate_pct),
        "is_vat_recoverable": a.is_vat_recoverable,
        "currency": a.currency,
        "exchange_rate": str(a.exchange_rate) if a.exchange_rate is not None else None,
        "supplier_id": str(a.supplier_id) if a.supplier_id else None,
        "supplier_name_snapshot": a.supplier_name_snapshot,
        "supplier_invoice_ref": a.supplier_invoice_ref,
        "is_cis_applicable": a.is_cis_applicable,
        "retention_released": a.retention_released,
        "linked_commitment_id": str(a.linked_commitment_id) if a.linked_commitment_id else None,
        "related_subcontract_id": str(a.related_subcontract_id) if a.related_subcontract_id else None,
        "is_reconciled_to_xero": a.is_reconciled_to_xero,
        "status": a.status,
        "posted_at": a.posted_at.isoformat() if a.posted_at else None,
        "paid_date": a.paid_date.isoformat() if a.paid_date else None,
        "payment_reference": a.payment_reference,
        "disputed_at": a.disputed_at.isoformat() if a.disputed_at else None,
        "voided_at": a.voided_at.isoformat() if a.voided_at else None,
        "created_at": a.created_at.isoformat() if a.created_at else None,
        "updated_at": a.updated_at.isoformat() if a.updated_at else None,
    }
    if include_sensitive:
        base.update({
            "cis_deduction_rate_pct": (
                str(a.cis_deduction_rate_pct) if a.cis_deduction_rate_pct is not None else None
            ),
            "cis_labour_amount": (
                str(a.cis_labour_amount) if a.cis_labour_amount is not None else None
            ),
            "cis_materials_amount": (
                str(a.cis_materials_amount) if a.cis_materials_amount is not None else None
            ),
            "cis_deduction_amount": (
                str(a.cis_deduction_amount) if a.cis_deduction_amount is not None else None
            ),
            "retention_rate_pct": (
                str(a.retention_rate_pct) if a.retention_rate_pct is not None else None
            ),
            "retention_amount": (
                str(a.retention_amount) if a.retention_amount is not None else None
            ),
            "retention_release_date": (
                a.retention_release_date.isoformat()
                if a.retention_release_date else None
            ),
            "reconciliation_variance": (
                str(a.reconciliation_variance) if a.reconciliation_variance is not None else None
            ),
            "dispute_reason": a.dispute_reason,
            "void_reason": a.void_reason,
            "ai_capture_metadata": a.ai_capture_metadata,
        })
    return base


def _serialise_change_log(c: ActualChangeLog) -> dict:
    return {
        "id": str(c.id),
        "actual_id": str(c.actual_id),
        "event_type": c.event_type,
        "actor_user_id": str(c.actor_user_id) if c.actor_user_id else None,
        "event_payload": c.event_payload,
        "occurred_at": c.occurred_at.isoformat() if c.occurred_at else None,
    }


def _serialise_attachment(att: ActualAttachment) -> dict:
    return {
        "id": str(att.id),
        "actual_id": str(att.actual_id),
        "original_filename": att.original_filename,
        "file_type": att.file_type,
        "file_size_bytes": att.file_size_bytes,
        "source": att.source,
        "uploaded_by_user_id": str(att.uploaded_by_user_id) if att.uploaded_by_user_id else None,
        "uploaded_at": att.uploaded_at.isoformat() if att.uploaded_at else None,
    }


def _include_sensitive(perms: UserPermissions) -> bool:
    return perms.has("actuals.view_sensitive") or perms.is_super_admin


# ---------------------------------------------------------------------
# 1. List (generic + project-scoped)
# ---------------------------------------------------------------------

@router.get("/actuals")
def list_actuals(
    filters: ActualsListFilters = Depends(_actuals_filters_dep),
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("actuals.view")),
    db: Session = Depends(get_db),
):
    rows, total = actuals_svc.list_actuals(
        db, filters=filters, user=current, perms=perms,
    )
    inc = _include_sensitive(perms)
    return {
        "items": [_serialise_actual(a, include_sensitive=inc) for a in rows],
        "count": len(rows),
        "total": total,
        "limit": filters.limit,
        "offset": filters.offset,
    }


@router.get("/projects/{project_id}/actuals")
def list_project_actuals(
    project_id: uuid.UUID,
    status: Optional[str] = Query(default=None),
    budget_line_id: Optional[uuid.UUID] = Query(default=None),
    limit: int = Query(default=100, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("actuals.view")),
    db: Session = Depends(get_db),
):
    # D32 — wrap mid-handler model construction so a `field_validator`
    # raise (e.g. invalid comma-separated status) surfaces as 422 rather
    # than 500. B34 closure.
    try:
        filters = ActualsListFilters(
            project_id=project_id, status=status,
            budget_line_id=budget_line_id,
            limit=limit, offset=offset,
        )
    except ValidationError as exc:
        raise HTTPException(
            status_code=422,
            detail=exc.errors(include_url=False, include_context=False),
        ) from exc
    rows, total = actuals_svc.list_actuals(
        db, filters=filters, user=current, perms=perms,
    )
    inc = _include_sensitive(perms)
    return {
        "project_id": str(project_id),
        "items": [_serialise_actual(a, include_sensitive=inc) for a in rows],
        "count": len(rows),
        "total": total,
    }


# ---------------------------------------------------------------------
# 2. Create + detail + update + delete
# ---------------------------------------------------------------------

@router.post("/actuals", status_code=201)
def create_actual(
    body: CreateActualRequest,
    request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("actuals.create")),
    db: Session = Depends(get_db),
):
    try:
        a = actuals_svc.create_actual(
            db, payload=body, user=current, perms=perms, request=request,
        )
    except ActualError as exc:
        _raise_for(exc)
    db.commit()
    db.refresh(a)
    return _serialise_actual(a, include_sensitive=_include_sensitive(perms))


@router.get("/actuals/{actual_id}")
def get_actual(
    actual_id: uuid.UUID,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("actuals.view")),
    db: Session = Depends(get_db),
):
    try:
        a = actuals_svc._load_actual(db, actual_id, current, perms)
    except ActualError as exc:
        _raise_for(exc)
    return _serialise_actual(a, include_sensitive=_include_sensitive(perms))


@router.patch("/actuals/{actual_id}")
def update_actual(
    actual_id: uuid.UUID,
    body: UpdateDraftActualRequest,
    request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("actuals.edit")),
    db: Session = Depends(get_db),
):
    try:
        a = actuals_svc.update_draft_actual(
            db, actual_id=actual_id, payload=body,
            user=current, perms=perms, request=request,
        )
    except ActualError as exc:
        _raise_for(exc)
    db.commit()
    db.refresh(a)
    return _serialise_actual(a, include_sensitive=_include_sensitive(perms))


@router.delete("/actuals/{actual_id}", status_code=204)
def delete_actual(
    actual_id: uuid.UUID,
    request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("actuals.edit")),
    db: Session = Depends(get_db),
):
    try:
        actuals_svc.delete_draft_actual(
            db, actual_id=actual_id, user=current, perms=perms, request=request,
        )
    except ActualError as exc:
        _raise_for(exc)
    db.commit()
    return Response(status_code=204)


@router.get("/actuals/{actual_id}/change-log")
def get_change_log(
    actual_id: uuid.UUID,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("actuals.view")),
    db: Session = Depends(get_db),
):
    try:
        rows = actuals_svc.get_change_log(
            db, actual_id=actual_id, user=current, perms=perms,
        )
    except ActualError as exc:
        _raise_for(exc)
    return {
        "actual_id": str(actual_id),
        "items": [_serialise_change_log(c) for c in rows],
        "count": len(rows),
    }


# ---------------------------------------------------------------------
# 3. State transitions
# ---------------------------------------------------------------------

@router.post("/actuals/{actual_id}/post")
def post_actual(
    actual_id: uuid.UUID,
    body: PostActualRequest,
    request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("actuals.edit")),
    db: Session = Depends(get_db),
):
    try:
        a = actuals_svc.post_actual(
            db, actual_id=actual_id, user=current, perms=perms,
            request=request, notes=body.notes,
        )
    except ActualError as exc:
        _raise_for(exc)
    db.commit()
    db.refresh(a)
    return _serialise_actual(a, include_sensitive=_include_sensitive(perms))


@router.post("/actuals/{actual_id}/mark-paid")
def mark_paid(
    actual_id: uuid.UUID,
    body: MarkPaidRequest,
    request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("actuals.approve")),
    db: Session = Depends(get_db),
):
    try:
        a = actuals_svc.mark_paid(
            db, actual_id=actual_id, paid_date=body.paid_date,
            payment_reference=body.payment_reference,
            user=current, perms=perms, request=request,
        )
    except ActualError as exc:
        _raise_for(exc)
    db.commit()
    db.refresh(a)
    return _serialise_actual(a, include_sensitive=_include_sensitive(perms))


@router.post("/actuals/{actual_id}/void")
def void_actual(
    actual_id: uuid.UUID,
    body: VoidActualRequest,
    request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("actuals.approve")),
    db: Session = Depends(get_db),
):
    try:
        a = actuals_svc.void_actual(
            db, actual_id=actual_id, void_reason=body.void_reason,
            user=current, perms=perms, request=request,
        )
    except ActualError as exc:
        _raise_for(exc)
    db.commit()
    db.refresh(a)
    return _serialise_actual(a, include_sensitive=_include_sensitive(perms))


@router.post("/actuals/{actual_id}/dispute")
def dispute_actual(
    actual_id: uuid.UUID,
    body: DisputeActualRequest,
    request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("actuals.edit")),
    db: Session = Depends(get_db),
):
    try:
        a = actuals_svc.dispute_actual(
            db, actual_id=actual_id, dispute_reason=body.dispute_reason,
            user=current, perms=perms, request=request,
        )
    except ActualError as exc:
        _raise_for(exc)
    db.commit()
    db.refresh(a)
    return _serialise_actual(a, include_sensitive=_include_sensitive(perms))


@router.post("/actuals/{actual_id}/undispute")
def undispute_actual(
    actual_id: uuid.UUID,
    body: UndisputeActualRequest,
    request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("actuals.edit")),
    db: Session = Depends(get_db),
):
    try:
        a = actuals_svc.undispute_actual(
            db, actual_id=actual_id, user=current, perms=perms,
            request=request, notes=body.notes,
        )
    except ActualError as exc:
        _raise_for(exc)
    db.commit()
    db.refresh(a)
    return _serialise_actual(a, include_sensitive=_include_sensitive(perms))


@router.post("/actuals/{actual_id}/release-retention")
def release_retention(
    actual_id: uuid.UUID,
    body: ReleaseRetentionRequest,
    request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("actuals.approve")),
    db: Session = Depends(get_db),
):
    try:
        a = actuals_svc.release_retention(
            db, actual_id=actual_id,
            retention_release_date=body.retention_release_date,
            user=current, perms=perms, request=request,
        )
    except ActualError as exc:
        _raise_for(exc)
    db.commit()
    db.refresh(a)
    return _serialise_actual(a, include_sensitive=_include_sensitive(perms))


# ---------------------------------------------------------------------
# 4. Attachments
# ---------------------------------------------------------------------

@router.get("/actuals/{actual_id}/attachments")
def list_attachments(
    actual_id: uuid.UUID,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("actuals.view")),
    db: Session = Depends(get_db),
):
    try:
        rows = att_svc.list_attachments(
            db, actual_id=actual_id, user=current, perms=perms,
        )
    except ActualError as exc:
        _raise_for(exc)
    return {
        "actual_id": str(actual_id),
        "items": [_serialise_attachment(a) for a in rows],
        "count": len(rows),
    }


@router.post("/actuals/{actual_id}/attachments", status_code=201)
def upload_attachment(
    actual_id: uuid.UUID,
    request: Request,
    file: UploadFile = File(...),
    source: str = Form(default="Manual_Upload"),
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("actuals.edit")),
    db: Session = Depends(get_db),
):
    if source not in ("Manual_Upload", "Email_Capture", "AI_Capture"):
        raise HTTPException(400, "Invalid source")
    try:
        att = att_svc.add_attachment(
            db, actual_id=actual_id,
            file_stream=file.file,
            original_filename=file.filename or "unnamed",
            content_type=file.content_type or "application/octet-stream",
            source=source,
            user=current, perms=perms, request=request,
        )
    except ActualError as exc:
        _raise_for(exc)
    db.commit()
    return _serialise_attachment(att)


@router.delete("/actuals/{actual_id}/attachments/{attachment_id}", status_code=204)
def delete_attachment(
    actual_id: uuid.UUID,
    attachment_id: uuid.UUID,
    request: Request,
    current: User = Depends(get_current_user),
    perms: UserPermissions = Depends(require_permission("actuals.edit")),
    db: Session = Depends(get_db),
):
    try:
        att_svc.remove_attachment(
            db, actual_id=actual_id, attachment_id=attachment_id,
            user=current, perms=perms, request=request,
        )
    except ActualError as exc:
        _raise_for(exc)
    db.commit()
    return Response(status_code=204)
